File: Data Structure/6. Queue/linkedListQueue.py
# Node class
class Node:
  # constructor
  def __init__(self, data):
    self.data = data
    self.next = None

# a linked list implementation of queue
class LinkedListQueue:
  # constructor
  def __init__(self):
    self.head = None
    self.length = 0

  # check if the queue is empty
  def is_empty(self):
    return self.length == 0

  # return the size of the queue
  def size(self):
    return self.length

  # peek the front of the queue
  def front(self):
    # if empty
    if self.is_empty():
      print("The queue is empty! No content in queue")
    else:
      return self.head.data

  # enqueue value to the top of the queue
  def enqueue(self, value):
    temp = Node(value) # create a new node
    
    # first check if the list is empty
    if self.is_empty():
      self.head = temp # let temp be the first node
      self.length += 1 # increment the size of the queue
      return
    else:
      # not empty, iterate and insert back
      current = self.head 
      while current.next != None:
        current = current.next # go to next node until None
      current.next = temp # let temp be the new last node
      self.length += 1 # increment the size of the queue
  
  # dequeue value from the queue top
  def dequeue(self):
    # if the queue if empty
    if self.is_empty():
      print("queue is empty! No need for dequeue it Now!")
    else:
      temp = self.head
      self.head = self.head.next # let new head be the next node
      temp.next = None # let the old head points to nothing
      self.length -= 1 # decrement the length of the queue
  
  # output the queue
  def output(self):
    # if empty
    if self.is_empty():
      print("No contents in the queue!")
    else:
      print("Contents: "),
      current = self.head
      while current != None:
        print(str(current.data) + " "),
        current = current.next # go to the next node until None
      print

# ============================
# testing the queue class
# ============================

print("\nLinked List implementation of queue!\n")

# create a new LinkedListQueue object
queue = LinkedListQueue()
print("- New ArrayStack object created")
queue.output()

# enqueue value into the queue
queue.enqueue(1)
queue.enqueue(2)
queue.enqueue(3)
print("\n- New values are pushed into the queue")
queue.output()

# peek the front of the queue
print("\n- Peek the front of the queue: " + str(queue.front()))
print

# dequeue the front of the queue
queue.dequeue()
print("- dequeue the front of the queue")
queue.output()

# get size of the queue
print("\n- Size of the queue: " + str(queue.size()))
print

File: Data Structure/6. Queue/Regular Queue/linkedListQueue.cpp
#include <iostream>
using namespace std;

// linked list implementation of queue
class linkedListQueue {
  struct Node {
    int data; // values in queue
    Node* next; // the next pointer
    // node constructor
    Node(int value): data(value), next(NULL) {}
  };
  
  Node* head; // the head pointer
  int length; // track the number of nodes
  
  public:
  // default constructor
  linkedListQueue() {
    head = NULL;
    length = 0;
  }

  // check if the queue is empty
  bool isEmpty() { return length == 0; }

   // return the size of the queue
  int size() { return length; }

  // get the front of the queue
  int front() { 
    if (isEmpty())
      cout << "queue is empty! No content in queue!" << endl;
    else
      return head -> data; 
  }

  // enqueue the new value to the back of the queue
  void enqueue(int value) {
    // create an temporary object
    Node* temp = new Node(value); // create a temporary node

    // first check if list is empty
    if (isEmpty()) {
      head = temp; // let temp be the first node
      length++;
      return;
    }

    // not empty list, iterate and insert back
    Node* current = head;
    while (current -> next != NULL) // go the last node
      current = current -> next; // go to next node
    current -> next = temp; // let temp be the last node
    length++; // increament the queue size
  }; 

  // pop the top element
  void dequeue() {
    // check if the queue is empty
    if (isEmpty()) {
      cout << "queue is empty! No need for deletion!" << endl;
    }
    // not empty, pop the top of the queue
    else {
      Node* temp = head; // get a copy of the head
      head = head -> next; // let the head point to the next node
      temp -> next = NULL; // let original head points to NULL 
      delete temp; // delete the top
      length--;
    }
  }

  // output the entire queue
  void output() {
    if (isEmpty()) {
      cout << "No Contents in the queue!" << endl;
    }
    else {
      cout << "Contents: ";
      Node* current = head;
      while (current != NULL) {
        cout << current -> data << " ";
        current = current -> next;
      }
      cout << endl;
    }
  }

  // clear the queue
  void clear() {
    if (isEmpty()) {
      cout << "List is empty! No need for deletion!" << endl;
    }
    // else the list is not empty, delete head
    else {
      // while queue is not empty, pop it
      while (!isEmpty()) 
        dequeue();
    } 
  }
};

// print function
void print(string text) {cout << text;}

// testing the linked list version of queue
int main() {
  print("Linked List implementation of queue\n\n");

  // create a queue
  linkedListQueue queue;
  print("- queue created!\n");
  queue.output(); // output if content exists
  print("\n");

  // enqueue value into the queue
  queue.enqueue(1);
  queue.enqueue(2);
  queue.enqueue(3);
  print("-values are pushed into the queue!\n");
  queue.output(); // output if content exists
  print("\n");

  // pop value from the top
  queue.dequeue();
  print("-pop the queue!\n");
  queue.output(); // output if content exists
  print("\n");

  // front of the queue
  cout << "-The front of the queue is: " << queue.front() << "\n\n";

  // size of the queue
  cout << "-The size is " << queue.size() << "\n";
  queue.output(); // output if content exists
  print("\n");
}